### PR TITLE
Reply Sheet UI

### PR DESCRIPTION
**Pull request by Sjmarf**
Added a non-functional reply sheet UI for posts (not comments yet). It includes a basic markdown editor, with buttons for the inline markdown elements, quotes and spoilers implemented. Later on, I'll add buttons for bulleted lists, numbered lists, headings, image uploading etc.

 In v1, you could swipe the sheet down to a lower detent so that you can interact with the content behind it. I've re-implemented this here, but the sheet goes much further down this time:
 

This will only happen if you swipe down whilst the text-editor is non-empty. Otherwise, the sheet will be closed.

### CODE REVIEW

File: Mlem.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1949,8 +1977,8 @@
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/mlemgroup/MlemMiddleware.git";
 			requirement = {
-				kind = upToNextMinorVersion;
-				minimumVersion = 0.15.0;
+				branch = sjmarf/reply;
+				kind = branch;
```
Comment (EricBAndrews) [inline Mlem.xcodeproj/project.pbxproj]: Reminder

File: Mlem/App/Views/Pages/ResponseComposerView.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                MarkdownTextEditor(
+                    text: $text,
+                    prompt: "Start writing...",
+                    textView: textView
+                ) {
+                    MarkdownEditorToolbarView(textView: textView)
+                }
+                .frame(minWidth: 0, maxWidth: .infinity, minHeight: minTextEditorHeight, maxHeight: .infinity)
+                Divider()
+                    .padding(.vertical, 10)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Pages/ResponseComposerView.swift]:
```suggestion
                    .padding(.vertical, AppConstants.standardSpacing)
```

File: Mlem/App/Views/Shared/MarkdownEditorToolbarView.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                    textView.wrapSelectionWithDelimiters("`")
+                }
+                Divider()
+                Button("Quote", systemImage: "quote.opening") {
+                    textView.toggleQuoteAtCursor()
+                }
+                Button("Image", systemImage: "photo") {}
+                Button("Spoiler", systemImage: "eye") {
+                    textView.wrapSelectionWithSpoiler()
+                }
```
Comment (EricBAndrews) [inline Mlem/App/Views/Shared/MarkdownEditorToolbarView.swift]: *nit*

System image names should go in `Icons`, there's an off chance we'll want them elsewhere

File: Mlem/App/Views/Shared/MarkdownTextEditor.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ 
+    func makeCoordinator() -> Coordinator {
+        Coordinator(self)
+    }
+ 
+    func makeUIView(context: Context) -> UITextView {
+        textView.font = .preferredFont(forTextStyle: .body)
+        textView.textContainerInset = .init(top: 5, left: 10, bottom: 10, right: 10)
+        textView.delegate = context.coordinator
+        // textView.setContentHuggingPriority(.defaultLow, for: .vertical)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Shared/MarkdownTextEditor.swift]: *nit*

Dead code?

File: Mlem/App/Views/Shared/MarkdownTextEditor.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        self.content = content()
+    }
+ 
+    func makeCoordinator() -> Coordinator {
+        Coordinator(self)
+    }
+ 
+    func makeUIView(context: Context) -> UITextView {
+        textView.font = .preferredFont(forTextStyle: .body)
+        textView.textContainerInset = .init(top: 5, left: 10, bottom: 10, right: 10)
```
Comment (EricBAndrews) [inline Mlem/App/Views/Shared/MarkdownTextEditor.swift]: *nit*

```suggestion
        textView.textContainerInset = .init(top: AppConstants.halfSpacing, left: AppConstants.standardSpacing, bottom: AppConstants.standardSpacing, right: AppConstants.standardSpacing)
```

**Review by EricBAndrews — CHANGES_REQUESTED** [review]
Really impressive work!

I've found a couple little defects:

- Selecting text, then making it bold, then italic results in the text being simply italic (it looks like the italic detector picks up the outer \*s and un-italics it). Going italic -> bold works as expected.
- Text past the third line vanishes into limbo along with the cursor--I think it goes off to the right of the screen, but it's hard to tell.

**Comment by Sjmarf** [comment]
> Selecting text, then making it bold, then italic results in the text being simply italic (it looks like the italic detector picks up the outer *s and un-italics it). Going italic -> bold works as expected.

Nice catch. My "solution" to this is to make italics use the alternative `_` instead of `*`. I think this is cleaner anyways, and it's what GitHub uses. This issue also exists with strike-through (`~~`) and subscript (`~`), though that is a much less used combination. Fixing this properly is possible, but the logic could get messy and considering that people won't run into it often I think it's safe to leave it for now. I've opened an issue [here](https://github.com/mlemgroup/mlem/issues/1158) to keep track of this.

> Text past the third line vanishes into limbo along with the cursor--I think it goes off to the right of the screen, but it's hard to tell.

Fixed 👍

**Review by EricBAndrews — APPROVED** [review]
LGTM!

Sorry this review took so long, my cat chewed through my charging cable and I couldn't test on-device lmao

**Comment by Sjmarf** [comment]
No worries xD

**Pull request merged**